File: src/clase-06/python/command/command.py
from __future__ import annotations
from abc import ABC, abstractmethod
from typing import List

class Command(ABC):
    @abstractmethod
    def execute(self) -> None: ...

class Light:
    def __init__(self):
        self.state = "off"
    def on(self) -> None:
        self.state = "on"
        print("Luz encendida")
    def off(self) -> None:
        self.state = "off"
        print("Luz apagada")

class LightOnCommand(Command):
    def __init__(self, light: Light):
        self.light = light
    def execute(self) -> None:
        self.light.on()

class LightOffCommand(Command):
    def __init__(self, light: Light):
        self.light = light
    def execute(self) -> None:
        self.light.off()

class RemoteControl:
    def __init__(self):
        self.history: List[Command] = []
    def run(self, cmd: Command) -> None:
        cmd.execute()
        self.history.append(cmd)

if __name__ == "__main__":
    light = Light()
    remote = RemoteControl()
    remote.run(LightOnCommand(light))
    remote.run(LightOffCommand(light))

File: src/clase-06/typescript/command/command.ts
export interface Command { execute(): void; }

export class Light {
  public state: 'on'|'off' = 'off';
  on(){ this.state = 'on'; console.log('Luz encendida'); }
  off(){ this.state = 'off'; console.log('Luz apagada'); }
}

export class LightOnCommand implements Command {
  constructor(private light: Light) {}
  execute(){ this.light.on(); }
}
export class LightOffCommand implements Command {
  constructor(private light: Light) {}
  execute(){ this.light.off(); }
}

export class RemoteControl {
  private history: Command[] = [];
  run(c: Command){ c.execute(); this.history.push(c); }
  getHistory(){ return this.history.slice(); }
}

// Demo
if (require.main === module) {
  const light = new Light();
  const remote = new RemoteControl();
  remote.run(new LightOnCommand(light));
  remote.run(new LightOffCommand(light));
}
